### DISCUSSION TITLE
Is it possible to load userContent.css with PWAs?

### DISCUSSION THREAD

**Discussion by AlbusPercivalDumbledore** [filips123/PWAsForFirefox#105] [Q&A] [answered]
@filips123 I'm trying to style a website I've now installed as PWA with usercontent.css, but I'm unable to do the same. In stock firefox, userContent.css is placed inside the chrome folder in target profile. Where do I place userContent.css for PWAs I've created?

Also, I'd like to modify browser chrome to my liking as well. Can userChrome.css be used for the same?

**Comment by filips123** [comment, marked as answer]
PWA profiles are in

   - Windows: `%APPDATA%\FirefoxPWA\profiles\`
   - Linux: `$XDG_DATA_HOME/firefoxpwa/profiles/` or `$HOME/.local/share/firefoxpwa/profiles/`
   - macOS: `$HOME/Library/Application Support/firefoxpwa/profiles/`

The default profile is `00000000000000000000000000`, other profiles will also have similar identifiers. To access `about:config` to enable CSS modifications, you can press `F6` inside a PWA and then enter `about:config`.

`userChrome.css` may also work, but it may also conflict with some other modifications done by  PWAsForFirefox itself.

**Comment by AlbusPercivalDumbledore** [comment]
@filips123 Thanks a lot! 

Was an oversight on my part. I had enabled css modifications with a user.js file, yet wasn't copying over all the css files from existing firefox installation.  userChrome.css too works as intented, except when at conflict with existing mods, as you mentioned.

One more question, if you don't mind. Would it be possible to load some userchromejs scripts as well? (I've been using some from https://github.com/aminomancer/uc.css.js)

**Reply by AlbusPercivalDumbledore** [reply]
Also, I might have found a bug.

Firefox Nightly 97.01a
Windows 10

Steps to Reproduce: 

1. Install uBlock Origin. 
2. Backup uBO settings from any other firefox installation. 
3. Try to import in PWA uBO (Restore from File). 
4. uBO would prompt for confirming settings overwrite. PWA crashes on confirming it.

**Reply by filips123** [reply]
Interesting, this doesn't happen to me. I don't know why doing things in extension would crash the whole browser, so it may be a problem with Firefox. Maybe you can try to reinstall uBlock Origin or recreate PWA profile.

**Reply by filips123** [reply]
Loading userchromejs scripts is harder because PWAsForFirefox also uses userchromejs and has to register its own loader. In theory, you could edit `C:\Program Files\FirefoxPWA\userchrome\runtime\_autoconfig.cfg` to include [fx-autoconfig's loader](https://github.com/MrOtherGuy/fx-autoconfig/blob/master/program/config.js), but the problem is that this file will be overwritten on every PWAsForFirefox update. But I have one idea how I could improve this in the future.

**Reply by AlbusPercivalDumbledore** [reply]
Could be something to do with my own install then. I've created 3 PWA profiles, installed uBO on all of them, and each one of them crashed when overwriting uBO settings. Peculiarly, it doesn't happen when importing settings for other extensions. Doesn't even create a crash log. Here's a gif:

![enF2uS9MJi](https://user-images.githubusercontent.com/82368617/147393522-63f73e3a-6eb8-4b44-9f96-587ce9b69bae.gif)

`file will be overwritten on every PWAsForFirefox update. But I have one idea how I could improve this in the future`

That'd be great. 

Thanks, again!

**Reply by filips123** [reply]
I found why importing uBO settings crashes. It doesn't actually crash...

When uBO imports settings, it closes its tab. However, because in PWA there is only one tab opened, it will close the whole window. I think your settings should still be imported when you reopen PWA though. I agree this is weird, but considering settings are imported and there is not much I can do, I'll leave it as is.

**Reply by filips123** [reply]
In https://github.com/filips123/PWAsForFirefox/commit/fc348607ec661a501af83a977a333b45e293bdae, I added support for loading third-party UC scripts without changing main PWAsForFirefox loader. You will need to save fx-autoconfig's loader as `YOUR-PROFILE/chrome/user/boot.jsm` and it should be automatically loaded. This will be available in the next release.

**Reply by AlbusPercivalDumbledore** [reply]
> When uBO imports settings, it closes its tab. However, because in PWA there is only one tab opened, it will close the whole window.

Ah, that makes sense. Yeah, settings get imported alright.

Thanks a ton for script support!
